File: js-src/tetris.js
import { Board } from './board.js';
import { BoardView } from './boardView.js';
import { Console } from './console.js';
import { PieceFactory } from './pieceFactory.js';

/**
 * Main Tetris game logic and game loop.
 * Converted from Java Tetris class.
 */
export class Tetris {
  /**
   * Creates a new Tetris game instance.
   * @param {number} width - The board width
   * @param {number} height - The board height
   */
  constructor(width, height) {
    this.console = new Console();
    this.board = new Board(width, height);
    this.boardView = new BoardView(this.console);
    this.gameRunning = true;
    this.score = 0;
    this.currentPieceView = null;
  }

  /**
   * Spawns a new random piece.
   */
  spawnNewPiece() {
    const pieces = [
      PieceFactory.createIPiece(),
      PieceFactory.createOPiece(),
      PieceFactory.createTPiece(),
      PieceFactory.createLPiece()
    ];

    const pieceViews = [
      PieceFactory.createIPieceView(),
      PieceFactory.createOPieceView(),
      PieceFactory.createTPieceView(),
      PieceFactory.createLPieceView()
    ];

    const randomIndex = Math.floor(Math.random() * pieces.length);
    this.board.setCurrentPiece(pieces[randomIndex]);
    this.currentPieceView = pieceViews[randomIndex];
  }

  /**
   * Processes user input commands.
   * @param {string} input - The user input
   */
  processUserInput(input) {
    if (input.length !== 0) {
      const command = input.charAt(0);
      const currentPiece = this.board.getCurrentPiece();

      switch (command) {
        case '4': // Move left
          if (this.board.canMovePiece(currentPiece, -1, 0)) {
            currentPiece.moveLeft();
          }
          break;
        case '6': // Move right
          if (this.board.canMovePiece(currentPiece, 1, 0)) {
            currentPiece.moveRight();
          }
          break;
        case '7': // Rotate counter-clockwise
          this.board.canRotatePiece(currentPiece, false);
          break;
        case '9': // Rotate clockwise
          this.board.canRotatePiece(currentPiece, true);
          break;
      }
    }
  }

  /**
   * Main game loop.
   */
  async gameLoop() {
    this.spawnNewPiece();
    
    while (this.gameRunning) {
      this.boardView.display(this.board, this.currentPieceView);
      
      const input = await this.console.readString(
        `Score: ${this.score}\n\nComando (4=izq, 6=der, 7=rotar↺, 9=rotar↻): `
      );
      
      this.processUserInput(input);

      const currentPiece = this.board.getCurrentPiece();

      // Try to move piece down
      if (this.board.canMovePiece(currentPiece, 0, 1)) {
        currentPiece.moveDown();
      } else {
        // Piece has landed, place it permanently
        this.board.placePiece(currentPiece, this.currentPieceView.getSymbol());

        // Check for completed lines
        const linesCleared = this.board.clearCompleteLines();
        this.score += linesCleared * 100;

        if (linesCleared > 0) {
          this.console.writeln(`¡${linesCleared} línea(s) eliminada(s)!`);
          await this.console.readString('Presiona Enter para continuar...');
        }

        // Spawn new piece
        this.spawnNewPiece();

        // Check for game over
        if (!this.board.canMovePiece(this.board.getCurrentPiece(), 0, 0)) {
          this.gameRunning = false;
          this.console.writeln('¡GAME OVER!');
          this.console.writeln(`Puntuación final: ${this.score}`);
        }
      }
    }

    this.console.close();
  }
}

File: src/main/java/Tetris.java
public class Tetris {
    private Board board;
    private BoardView boardView;
    private boolean gameRunning;
    private int score;
    private Console console;
    private PieceView currentPieceView;

    public Tetris(int width, int height) {
        console = new Console();
        board = new Board(width, height);
        boardView = new BoardView(console);
        gameRunning = true;
        score = 0;
    }

    public void spawnNewPiece() {
        Piece[] pieces = {
                PieceFactory.createIPiece(),
                PieceFactory.createOPiece(),
                PieceFactory.createTPiece(),
                PieceFactory.createLPiece()
        };

        PieceView[] pieceViews = {
                PieceFactory.createIPieceView(),
                PieceFactory.createOPieceView(),
                PieceFactory.createTPieceView(),
                PieceFactory.createLPieceView()
        };

        int randomIndex = (int) (Math.random() * pieces.length);
        board.setCurrentPiece(pieces[randomIndex]);
        currentPieceView = pieceViews[randomIndex];
    }

    private void processUserInput(String input) {
        if (input.length() != 0) {
            char command = input.charAt(0);
            Piece currentPiece = board.getCurrentPiece();

            switch (command) {
                case '4' -> {
                    if (board.canMovePiece(currentPiece, -1, 0))
                        currentPiece.moveLeft();
                }
                case '6' -> {
                    if (board.canMovePiece(currentPiece, 1, 0))
                        currentPiece.moveRight();
                }
                case '7' -> board.canRotatePiece(currentPiece, false);
                case '9' -> board.canRotatePiece(currentPiece, true);
            }
        }
    }

    public void gameLoop() {
        spawnNewPiece();
        while (gameRunning) {
            boardView.display(board, currentPieceView);
            String input = console.readString("Score: " + score + "\n\nComando (4=izq, 6=der, 7=rotar↺, 9=rotar↻): ");
            processUserInput(input);

            Piece currentPiece = board.getCurrentPiece();

            if (board.canMovePiece(currentPiece, 0, 1)) {
                currentPiece.moveDown();
            } else {
                board.placePiece(currentPiece, currentPieceView.getSymbol());

                int linesCleared = board.clearCompleteLines();
                score += linesCleared * 100;

                if (linesCleared > 0) {
                    console.writeln("¡" + linesCleared + " línea(s) eliminada(s)!");
                    console.readString("Presiona Enter para continuar...");
                }

                spawnNewPiece();

                if (!board.canMovePiece(board.getCurrentPiece(), 0, 0)) {
                    gameRunning = false;
                    console.writeln("¡GAME OVER!");
                    console.writeln("Puntuación final: " + score);
                }
            }
        }
    }
}